cmd/src/update.go

```diff
@@ -0,0 +1,156 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+
+	"github.com/tj/go-update"
+	"github.com/tj/go-update/progress"
+	"github.com/tj/go-update/stores/github"
+)
+
+const (
+	// PrivateFileMode grants owner to read/write a file.
+	PrivateFileMode = 0600
+)
+
+func init() {
+	usage := `
+Examples:
+
+	Update and replace src-cli to the recommended version:
+
+		$ src update
+`
+
+	flagSet := flag.NewFlagSet("update", flag.ExitOnError)
+
+	var apiFlags = api.NewFlags(flagSet)
+
+	handler := func(args []string) error {
+		currentDirectory, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		err = isDirWritable(currentDirectory)
+		if err != nil {
+			runWithElevatedPrivilege()
+			return nil
+		}
+
+		fmt.Printf("Current version: %s\n", buildTag)
+
+		client := cfg.apiClient(apiFlags, flagSet.Output())
+		recommendedVersion, err := getRecommendedVersion(context.Background(), client)
+		if err != nil {
+			return err
+		}
+		if recommendedVersion == "" {
+			fmt.Println("Recommended Version: <unknown>")
+			return nil
+		}
+		if buildTag == recommendedVersion {
+			fmt.Printf("src-cli is already at the recommended version: %s\n", recommendedVersion)
+			return nil
+		}
+
+		// Retrieving latest release information.
+		updateManager := &update.Manager{
+			Command: "src",
+			Store: &github.Store{
+				Owner:   "sourcegraph",
+				Repo:    "src-cli",
+				Version: buildTag,
+			},
+		}
+		if runtime.GOOS == "windows" {
+			updateManager.Command += ".exe"
+		}
+		releases, err := updateManager.LatestReleases()
+		if err != nil {
+			return err
+		}
+		if len(releases) == 0 {
+			return fmt.Errorf("No latest src-cli release update")
+		}
+		latestRelease := releases[0]
+		if latestRelease.Version != recommendedVersion {
+			return fmt.Errorf("Mismatch of recommended version")
+		}
+
+		updateAsset := latestRelease.FindTarball(runtime.GOOS, runtime.GOARCH)
+		if updateAsset == nil {
+			return fmt.Errorf("src-cli binary has not been installed in this machine")
+		}
+		fmt.Printf("Downloading src-cli version %s\n", recommendedVersion)
+		tarball, err := updateAsset.DownloadProxy(progress.Reader)
```

Comment:
We could use our own `output` package here and save the dependency on `progress`. I just looked at what `progress.Reader` is and what `DownloadProxy` expects and it wouldn't need a lot of code.

Hacky version:

```suggestion
		out := output.NewOutput(flagSet.Output(), output.OutputOpts{})
		tarball, err := updateAsset.DownloadProxy(newProgressReader(out, "Downloading src-cli"))
```

And then somewhere else (ideally in the `output` package?)

```go
func newProgressReader(out *output.Output, message string) update.Proxy {
	return func(size int, r io.ReadCloser) io.ReadCloser {
		progress := out.Progress([]output.ProgressBar{{
			Label: message,
			Max:   float64(size),
		}}, nil)
		return &customThing{bars: progress, r: r}
	}
}

var _ io.ReadCloser = &customThing{}

type customThing struct {
	bars    output.Progress
	r       io.ReadCloser
	written int
}

func (c *customThing) Read(b []byte) (int, error) {
	n, err := c.r.Read(b)
	c.written += n
	c.bars.SetValue(0, float64(c.written))
	return n, err
}

func (c *customThing) Close() error {
	c.bars.Complete()
	return c.r.Close()
}
```

Works:


https://user-images.githubusercontent.com/1185253/106127224-9fc15600-615e-11eb-933d-225deffeac3f.mp4